### PR TITLE
nixd/completion: eval files to make completion list

### PR DESCRIPTION
**Pull request by inclyc**

### CODE REVIEW

File: lib/nixd/src/Server.cpp [hunk 1]
```diff
@@ -6,6 +6,7 @@
 #include "lspserver/Logger.h"
 #include "lspserver/Path.h"
 #include "lspserver/Protocol.h"
+#include "value.hh"
```
Comment (inclyc) [inline lib/nixd/src/Server.cpp:9]:
```suggestion
#include "<nix/value.hh>"
```

File: lib/nixd/include/nixd/AST.h [hunk 2]
```diff
@@ -47,7 +47,7 @@ class EvalAST {
   /// Get the corresponding 'Env' while evaluating the expression.
   /// nix 'Env's contains dynamic variable name bindings at evaluation, might be
   /// used for completion.
-  nix::Env getEnv(nix::Expr *Expr) { return EnvMap.at(Expr); }
+  const nix::Env *getEnv(nix::Expr *Expr) { return EnvMap.at(Expr); }
```
Comment (junaire) [inline lib/nixd/include/nixd/AST.h:50]: const correctness?

File: lib/nixd/src/Server.cpp [hunk 3]
```diff
@@ -160,6 +161,54 @@ void Server::addDocument(lspserver::PathRef File, llvm::StringRef Contents,
            [](std::shared_ptr<EvalDraftStore::EvalResult> Result) {});
 }
 
+CompletionHelper::Items
+CompletionHelper::fromStaticEnv(const nix::SymbolTable &STable,
+                                const nix::StaticEnv &SEnv) {
+  Items Result;
+  for (auto [Symbol, Displ] : SEnv.vars) {
```
Comment (junaire) [inline lib/nixd/src/Server.cpp:168]: `Symbol` is used as a ref while `Displ` is unused?

File: lib/nixd/src/Server.cpp [hunk 4]
```diff
@@ -160,6 +161,54 @@ void Server::addDocument(lspserver::PathRef File, llvm::StringRef Contents,
...
+  }
+  return Result;
+}
+
+CompletionHelper::Items
+CompletionHelper::fromEnvWith(const nix::SymbolTable &STable,
+                              const nix::Env &NixEnv) {
+  Items Result;
+  if (NixEnv.type == nix::Env::HasWithAttrs) {
+    nix::Bindings::iterator BindingIt = NixEnv.values[0]->attrs->begin();
```
Comment (junaire) [inline lib/nixd/src/Server.cpp:182]: auto

File: lib/nixd/include/nixd/AST.h (hunk 2 above)
Comment (inclyc) [reply lib/nixd/include/nixd/AST.h:50]: y

File: lib/nixd/src/Server.cpp (hunk 3 above)
Comment (inclyc) [reply lib/nixd/src/Server.cpp:168]: They can be trivially copied (actually they are integers)

File: lib/nixd/src/Server.cpp [hunk 5]
```diff
@@ -304,4 +354,57 @@ void Server::onHover(const lspserver::TextDocumentPositionParams &Paras,
...
+             CompletionHelper::Items Items;
+             try {
+               auto AST = Result->EvalASTForest.at(CompletionFile);
+               auto *Node = AST->lookupPosition(Params.position);
+
+               lspserver::vlog("current trigger character is {0}",
+                               Params.context.triggerCharacter);
+               if (Params.context.triggerCharacter == ".") {
+                 auto Value = AST->getValue(Node);
+                 std::stringstream SS;
```
Comment (inclyc) [inline lib/nixd/src/Server.cpp:384]: nit: unused stuff

File: lib/nixd/include/nixd/AST.h (hunk 2 above)
Comment (junaire) [inline lib/nixd/include/nixd/AST.h:50]:
```suggestion
  const nix::Env *getEnv(nix::Expr *Expr) const { return EnvMap.at(Expr); }
```

**Pull request merged**